wiki/en/JSON-RPC.md

```diff
@@ -0,0 +1,221 @@
+---
+layout: wiki
+title: "JSON-RPC interface"
+lang: "en"
+permalink: "/wiki/JSON-RPC"
+---
+
+# JSON-RPC interface
+
```

Comment:
As I assume this page will grow in time (and is already quite extensive), you may want to add a table of contents.

```suggestion
 {:.no_toc}

This manual documents the JSON-RPC-based protocol by which you can control the Jamulus client and server.

<details markdown="1">

<summary>Table of contents</summary>

* TOC
 {:toc}

</details> 
```

Reply:
@gilgongo I am thinking about whether we need an extensive documentation for JSON-RPC at all…

My concern is that this would double the work required to make changes to the JSON-RPC interface. First you would need to make changes in `jamulus` project, and then you would need to document the method here in `jamuluswebsite`.

I think that adds substantial maintenance burden. I want to avoid having to make changes to 2 repositories and filing 2 PRs. So, upon second thought, I am thinking of doing either of these:

- Point the reader to read `clientrpc.cpp` and `serverrpc.cpp`. Since the code is quite self-documenting already, it may be a redundant to have an English documentation at all.

- Add some code comments directly in `clientrpc.cpp` and `serverrpc.cpp` and create a Python script generate a Markdown documentation based on these comments (e.g. `tools/generate_rpc_docs.py`). The documentation would live directly in jamulus repository, e.g. `RPC.md`, and jamuluswebsite would just link there. Maybe some GitHub Actions could be created to ensure that this documentation is up-to-date. This would also lets reader view the documentation of a specific release.

  The comment might look like this:

  ```cpp
        /// @rpc_notification jamulusclient/clientListReceived
        /// @brief Emitted when the client list is received.
        /// @param {array} params.clients - The client list.
        /// @param {number} params.clients[*].id - The channel ID.
        /// @param {string} params.clients[*].name - The musician’s name.
        /// @param {string} params.clients[*].skillLevel - The musician’s skill level (beginner, intermediate, expert, or null).
        /// @param {number} params.clients[*].countryId - The musician’s country ID (see QLocale).
        /// @param {string} params.clients[*].city - The musician’s city.
        /// @param {number} params.clients[*].instrumentId - The musician’s instrument ID (see CInstPictures::GetTable).
  ````

Suggestions or opinions?

Reply:
I'm not qualified to say, but if it's self-documenting, then less work is always better/safer.

Reply:
Could we add a more extensive documentation to the file (md table like here and then link the comment from here)? I don’t think people should be forced to read the code. But they can of course read comments - although that’s not optimal (consistency, style) but ok for an experimental feature.

Reply:
We should also consider the audience of this documentation: It will be developers (not necessarily Jamulus developers or C++ developers). They should already be familiar with reading code or at least comments in the code. I like the idea of having a single place to handle both the code and the directly related docs.

I think we should still keep the first part in the "wiki" (everything including the **Example** paragraph?), put the method documentation right next to the code and link from the wiki page to the code.